infrastructure/local-provisioners/azurerm_apim.ts

```diff
@@ -0,0 +1,510 @@
+/**
+ * Run this task to deploy Azure API Manager:
+ *
+ * yarn resources:apim:setup
+ *
+ * This task assumes that the following resources are already created:
+ *  - Resource group
+ *  - Functions (app service)
+ *
+ * Unfortunately you cannot migrate Widgets and Media Libray:
+ * https://{publisherPortalName}.portal.azure-api.net/Admin/Widgets
+ */
+// tslint:disable:no-console
+// tslint:disable:no-any
+
+import * as t from "io-ts";
+
+import { reporter } from "io-ts-reporters";
+
+import * as winston from "winston";
+
+import yargs = require("yargs");
+
+import { ICreds, login } from "../../lib/login";
+import { getFunctionsInfo } from "../../lib/task_utils";
+
+import { checkEnvironment } from "../../lib/environment";
+
+import apiManagementClient = require("azure-arm-apimanagement");
+import webSiteManagementClient = require("azure-arm-website");
+
+import * as fs from "fs";
+import * as path from "path";
+import * as shelljs from "shelljs";
+import * as tmp from "tmp";
+import * as url from "url";
+
+import { left } from "fp-ts/lib/Either";
+import * as replaceInFiles from "replace-in-file";
+import {
+  getObjectFromJson,
+  IResourcesConfiguration,
+  readConfig
+} from "../../lib/config";
+
+// tslint:disable-next-line:no-object-mutation
+shelljs.config.fatal = true;
+// shelljs.config.verbose = true;
+
+const CONFIGURATION_DIRECTORY_NAME = "apim";
```

Comment:
this config path (in general all parameters) should come from terraform (i.e. as command line parameter)

terraform variables should provide all configuration

Reply:
The API manager task needs a lot of variables, from his own configuration and from TF configuration as well. My first try was to pass them one by one in the command line which not only was unpractical, but it even does not work on win32 because "command line is too long". So I reverted the old behavior and now most of the parameters are read directly from the configuration files. Of course we can pass the configuration file path from command line (at least, of apim.json) but as there values are shared there would be no practical usage.

I've kept only the eventhub_connection string in the command line because it comes from TF eventhub resource. There are no other possible values which can be took from output vars of TF resources.

Reply:
I'm just suggesting to avoid having defaults hardcoded in the provisioner and making explicit that the provisioner is getting additional config from a file that not currently evident by reading the Terraform configuration (you can think about it as a kind of side effect)

Reply:
Moreover, the APIm configuration task replace all the placeholders it found in the developer portal template files with the actual values found in configuration files (which we must pass at whole).

Reply:
(ok)